// @salesforce/lightning-lsp-common/lib/html-language-service/htmlLanguageService.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getLanguageService = void 0;
const htmlScanner_1 = require("./parser/htmlScanner");
const htmlParser_1 = require("./parser/htmlParser");
const htmlCompletion_1 = require("./services/htmlCompletion");
const tagProviders_1 = require("./services/tagProviders");
const htmlHover_1 = require("./services/htmlHover");
const htmlFormatter_1 = require("./services/htmlFormatter");
const htmlLinks_1 = require("./services/htmlLinks");
const htmlHighlighting_1 = require("./services/htmlHighlighting");
const htmlSymbolsProvider_1 = require("./services/htmlSymbolsProvider");
const htmlDefinition_1 = require("./services/htmlDefinition");
const lwcExtensions_1 = require("./services/lwcExtensions");
const auraExtensions_1 = require("./services/auraExtensions");
const htmlFolding_1 = require("./services/htmlFolding");
__exportStar(require("./htmlLanguageTypes"), exports);
__exportStar(require("vscode-languageserver-types"), exports);
function getLanguageService() {
    const htmlCompletion = new htmlCompletion_1.HTMLCompletion();
    return {
        createScanner: htmlScanner_1.createScanner,
        addTagProvider: tagProviders_1.addTagProvider,
        getTagProviders: tagProviders_1.getTagProviders,
        parseHTMLDocument: document => (0, htmlParser_1.parse)(document.getText()),
        doComplete: htmlCompletion.doComplete.bind(htmlCompletion),
        setCompletionParticipants: htmlCompletion.setCompletionParticipants.bind(htmlCompletion),
        doHover: htmlHover_1.doHover,
        format: htmlFormatter_1.format,
        findDocumentHighlights: htmlHighlighting_1.findDocumentHighlights,
        findDocumentLinks: htmlLinks_1.findDocumentLinks,
        findDocumentSymbols: htmlSymbolsProvider_1.findDocumentSymbols,
        getFoldingRanges: htmlFolding_1.getFoldingRanges,
        doTagComplete: htmlCompletion.doTagComplete.bind(htmlCompletion),
        findDefinition: htmlDefinition_1.findDefinition,
        getPropertyBindingTemplateDeclaration: lwcExtensions_1.getPropertyBindingTemplateDeclaration,
        getPropertyBindingValue: lwcExtensions_1.getPropertyBindingValue,
        getAuraBindingTemplateDeclaration: auraExtensions_1.getAuraBindingTemplateDeclaration,
        getAuraBindingValue: auraExtensions_1.getAuraBindingValue,
    };
}
exports.getLanguageService = getLanguageService;
