// @salesforce/lightning-lsp-common/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Node = exports.parse = exports.getLanguageService = exports.interceptConsoleLogger = exports.MemberType = exports.Decorator = exports.AttributeInfo = exports.TagInfo = exports.shared = exports.componentUtil = exports.utils = exports.WorkspaceContext = void 0;
const utils = __importStar(require("./utils"));
exports.utils = utils;
const context_1 = require("./context");
Object.defineProperty(exports, "WorkspaceContext", { enumerable: true, get: function () { return context_1.WorkspaceContext; } });
const shared = __importStar(require("./shared"));
exports.shared = shared;
const tagInfo_1 = require("./indexer/tagInfo");
Object.defineProperty(exports, "TagInfo", { enumerable: true, get: function () { return tagInfo_1.TagInfo; } });
const attributeInfo_1 = require("./indexer/attributeInfo");
Object.defineProperty(exports, "AttributeInfo", { enumerable: true, get: function () { return attributeInfo_1.AttributeInfo; } });
Object.defineProperty(exports, "Decorator", { enumerable: true, get: function () { return attributeInfo_1.Decorator; } });
Object.defineProperty(exports, "MemberType", { enumerable: true, get: function () { return attributeInfo_1.MemberType; } });
const logger_1 = require("./logger");
Object.defineProperty(exports, "interceptConsoleLogger", { enumerable: true, get: function () { return logger_1.interceptConsoleLogger; } });
const componentUtil = __importStar(require("./component-util"));
exports.componentUtil = componentUtil;
const htmlLanguageService_1 = require("./html-language-service/htmlLanguageService");
Object.defineProperty(exports, "getLanguageService", { enumerable: true, get: function () { return htmlLanguageService_1.getLanguageService; } });
const htmlParser_1 = require("./html-language-service/parser/htmlParser");
Object.defineProperty(exports, "parse", { enumerable: true, get: function () { return htmlParser_1.parse; } });
Object.defineProperty(exports, "Node", { enumerable: true, get: function () { return htmlParser_1.Node; } });
